### PR TITLE
[PUBDEV-4292][WIP] Non-recursive creation of Rapid expressions in Python

### PR DESCRIPTION
**Pull request by mmalohlava**
The fix includes:
  - switch from recursive preparation of Rapids expresion to explicit stack-based preparation
  - optimized string concatenation (based on http://blog.mclemon.io/ironpython-efficient-string-concatenation)

TODO: backend is still failing to StackOverflow

### CODE REVIEW

**Comment by mmalohlava** [comment]
Note: This is still WIP - i need to change backend since it is using recursive descent parser as well :/

File: h2o-py/h2o/expr.py [hunk 1]
```diff
@@ -73,7 +72,7 @@ class ExprNode(object):
 
     # Magical count-of-5:   (get 2 more when looking at it in debug mode)
     #  2 for _get_ast_str frame, 2 for _get_ast_str local dictionary list, 1 for parent
-    MAGIC_REF_COUNT = 5 if sys.gettrace() is None else 7  # M = debug ? 7 : 5
+    MAGIC_REF_COUNT = 4 if sys.gettrace() is None else 6  # M = debug ? 6 : 4
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: Could we explain this a bit better? Also "Magical count-of-5:" should be maybe changed to "Magical count-of-4:"

File: h2o-py/h2o/expr.py [hunk 2]
```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
     # the timezone change, so cannot be lazy.
     #
     def _get_ast_str(self, top):
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: The comment should be changed to reflect the change of computation not being recursive anymore.

I would also document the meaning of `top` argument as I couldn't find the documentation anywhere. Does it mean that the frame should be cached ?

File: h2o-py/h2o/expr.py [hunk 3]
```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
...
-        return exec_str
+        stack = collections.deque()
+        # Append self and evaluate
+        stack.append(self)
+        r = collections.deque()
+        while len(stack) > 0:
+            head = stack.pop()
+            s = None
+            if isinstance(head, ExprNode):
+                if not head._cache.is_empty():  # Data already computed and cached; could a "false-like" cached value
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: Not sure what is meant by "could a "false-like" cached value. Could we explain it maybe a bit better for someone seeing this code for the first time?

File: h2o-py/h2o/expr.py [hunk 4]
```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
...
+                    s = str(head._cache._data) if head._cache.is_scalar() else head._cache._id
+                elif head._cache._id is not None:
+                    s = head._cache._id  # Data already computed under ID, but not cached
+                else:
+                    s = "({}".format(head._op)
+                    gc_ref_cnt = len(gc.get_referrers(head))
+                    if (top and head == self) or gc_ref_cnt >= ExprNode.MAGIC_REF_COUNT:
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: Does this step mean that if `top` is true ( -> cache it) and we are on the top-level frame, do cache?

File: h2o-py/h2o/expr.py [hunk 5]
```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
...
+                        head._cache._id = _py_tmp_key(append=h2o.connection().session_id)
+                        s = "(tmp= {} {}".format(head._cache._id, s)
+                        stack.append('^@^')
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: This is some special rapid expression '^@^' ?

File: h2o-py/h2o/expr.py [hunk 6]
```diff
@@ -121,24 +120,44 @@ def _eval_driver(self, top):
...
+                s = ')'
+            else:
+                s = ExprNode._arg_to_expr(head)
+            # Append
+            r.append(s)
+
+        return " ".join(r)
 
     @staticmethod
     def _arg_to_expr(arg):
```
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: Maybe rename this method to `primitive_as_string` or something similar?

File: h2o-py/h2o/expr.py (hunk 3 above)
Comment (mmalohlava) [inline h2o-py/h2o/expr.py]: ok

File: h2o-py/h2o/expr.py (hunk 4 above)
Comment (mmalohlava) [inline h2o-py/h2o/expr.py]: yes, my original idea was to remove top flag totally and use only position in stack, however some of calls passing explicit top param

File: h2o-py/h2o/expr.py (hunk 5 above)
Comment (mmalohlava) [inline h2o-py/h2o/expr.py]: 🍡 :)

Comment (mmalohlava) [inline h2o-py/h2o/expr.py]: it is just a marker on stack, but i should make it more clever.

File: h2o-py/h2o/expr.py (hunk 6 above)
Comment (mmalohlava) [inline h2o-py/h2o/expr.py]: good point!

File: h2o-py/h2o/expr.py (hunk 5 above)
Comment (jakubhava) [inline h2o-py/h2o/expr.py]: Cool, now I understand :) Nice trick!  would just but there a small comment explaining this, just for future

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
Still needs change of the backend, postponing for the next release.

**Comment by mmalohlava** [comment]
@h2o-ops please test!

File: h2o-py/h2o/expr.py [hunk 7]
```diff
@@ -135,24 +136,45 @@ def _optimize(self):
...
+            head = stack.pop()
+            s = None
+            if isinstance(head, ExprNode):
+                if not head._cache.is_empty():  # Data already computed and cached; could a "false-like" cached value
+                    s = str(head._cache._data) if head._cache.is_scalar() else head._cache._id
+                elif head._cache._id is not None:
+                    s = head._cache._id  # Data already computed under ID, but not cached
+                else:
+                    s = "({}".format(head._op)
+                    gc_ref_cnt = len(gc.get_referrers(head))
```
Comment (st-pasha) [inline h2o-py/h2o/expr.py]: would this be the same as `sys.getrefcount(head)` ?
Getting the refcount is simply accessing the `refcnt` field on an underlying `PyObject`; however collecting the list of referrers -- I don't know what it involves, making an entire gc sweep?

File: h2o-py/h2o/expr.py [hunk 8]
```diff
@@ -135,24 +136,45 @@ def _optimize(self):
...
+                    print(gc.get_referrers(head))
```
Comment (st-pasha) [inline h2o-py/h2o/expr.py]: Leftover debugging code

File: h2o-py/h2o/expr.py (hunk 1 above)
Comment (st-pasha) [reply h2o-py/h2o/expr.py]: This MAGIC_REF_COUNT thing is pure evil enfleshed. Once upon a time I spent several days debugging problems caused by it.
Ideally there should be a *clear* and *detailed* explanation of what this does and what the "magical" counts are. I suspect current comment are out-of-sync with reality...

File: h2o-py/tests/testdir_algos/gbm/pyunit_weights_gamma_gbm.py [hunk 9]
```diff
@@ -29,6 +29,7 @@ def weights_gamma():
 
 
 if __name__ == "__main__":
+  h2o.enable_expr_optimizations(False)
```
Comment (st-pasha) [inline h2o-py/tests/testdir_algos/gbm/pyunit_weights_gamma_gbm.py]: why is this needed? will the code not work without?

File: h2o-py/h2o/expr.py (hunk 7 above)
Comment (mmalohlava) [reply h2o-py/h2o/expr.py]: good point, thx!

File: h2o-py/h2o/expr.py (hunk 1 above)
Comment (mmalohlava) [reply h2o-py/h2o/expr.py]: agree @st-pasha , there is no well defined semantics of that field. I would like to replace that but slightly different implementation.
